File: backend/src/main/java/com/designtemplate/api/service/CategoryService.java
package com.designtemplate.api.service;

import com.designtemplate.api.dto.CategoryDto;
import com.designtemplate.api.model.Category;
import com.designtemplate.api.repository.CategoryRepository;
import com.designtemplate.api.repository.TemplateRepository;
import com.designtemplate.api.exception.ResourceNotFoundException;
import com.designtemplate.api.exception.InvalidOperationException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.security.core.context.SecurityContextHolder;

import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class CategoryService {
    private final CategoryRepository categoryRepository;
    private final TemplateRepository templateRepository;

    public List<CategoryDto> getAllCategories() {
        return categoryRepository.findAll().stream()
                .map(this::convertToDto)
                .collect(Collectors.toList());
    }

    public CategoryDto createCategory(CategoryDto categoryDto) {
        if (categoryRepository.existsByName(categoryDto.getName())) {
            throw new IllegalArgumentException("Category with this name already exists");
        }

        Category category = Category.builder()
                .name(categoryDto.getName())
                .description(categoryDto.getDescription())
                .createdBy(getCurrentUserId())
                .createdAt(LocalDateTime.now())
                .updatedAt(LocalDateTime.now())
                .build();

        Category savedCategory = categoryRepository.save(category);
        return convertToDto(savedCategory);
    }

    public void deleteCategory(String id) {
        Category category = categoryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found"));

        // Check if category is in use
        if (templateRepository.existsByCategory(category.getName())) {
            throw new InvalidOperationException("Cannot delete category that is being used by templates");
        }

        categoryRepository.deleteById(id);
    }

    public long getCategoryUsageCount(String categoryName) {
        return templateRepository.countByCategory(categoryName);
    }

    private String getCurrentUserId() {
        return SecurityContextHolder.getContext().getAuthentication().getName();
    }

    private CategoryDto convertToDto(Category category) {
        return CategoryDto.builder()
                .id(category.getId())
                .name(category.getName())
                .description(category.getDescription())
                .createdBy(category.getCreatedBy())
                .createdAt(category.getCreatedAt())
                .updatedAt(category.getUpdatedAt())
                .build();
    }
}

File: src/services/categoryService.ts
import api from "@/lib/api";

export interface Category {
  id: string;
  name: string;
  description: string;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
}

const categoryService = {
  getAllCategories: async (): Promise<Category[]> => {
    try {
      const response = await api.get<Category[]>('/api/categories');
      return response.data;
    } catch (error) {
      // Return default categories if API fails
      return [
        { id: '1', name: 'UI Design', description: 'User Interface Design', createdBy: '', createdAt: '', updatedAt: '' },
        { id: '2', name: 'UX Design', description: 'User Experience Design', createdBy: '', createdAt: '', updatedAt: '' },
        { id: '3', name: 'System Design', description: 'System Architecture Design', createdBy: '', createdAt: '', updatedAt: '' },
        { id: '4', name: 'API Design', description: 'API Design and Documentation', createdBy: '', createdAt: '', updatedAt: '' }
      ];
    }
  },

  createCategory: async (category: Partial<Category>) => {
    const response = await api.post('/categories', category);
    return response.data;
  },

  deleteCategory: async (id: string) => {
    await api.delete(`/categories/${id}`);
  },

  getCategoryUsageCount: async (categoryName: string): Promise<number> => {
    const response = await api.get(`/categories/${categoryName}/usage`);
    return response.data;
  }
};

export default categoryService;
